[JIT] Fix python pickle serialization for torchbind

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32938 [WIP][quantization] Use torchbind for Linear PackedParams
* #32977 [JIT] Resolve custom classes in source importer
* #32948 [c10] Allow taking a std::tuple as arg
* **#32878 [JIT] Fix python pickle serialization for torchbind**
* #32854 [JIT] Make IRParser use op schema
* #32833 [JIT] Trace uses of torchbind classes as module attributes

Differential Revision: [D19669879](https://our.internmc.facebook.com/intern/diff/D19669879)

Commit:
[JIT] Fix python pickle serialization for torchbind

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 074d2416:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=074d24160ea036a2588a961e7f62f63b86869e81).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4392100) pytorch_linux_xenial_py2_7_9_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93884663) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93884663))
<details>
<summary>
<code>Jan 31 21:59:04 /var/lib/jenkins/workspace/c10/util/Exception.h:215:29: error: nonnull argument 'this' compared to NULL [-Werror=nonnull-compare]</code>
</summary>

```
Jan 31 21:59:02 [ 81%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/distributed/autograd/utils.cpp.o 
Jan 31 21:59:03 [ 81%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/distributed/rpc/message.cpp.o 
Jan 31 21:59:03 [ 81%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/distributed/rpc/python_call.cpp.o 
Jan 31 21:59:03 [ 81%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/distributed/rpc/python_remote_call.cpp.o 
Jan 31 21:59:04 In file included from /var/lib/jenkins/workspace/c10/core/DeviceType.h:8:0, 
Jan 31 21:59:04                  from /var/lib/jenkins/workspace/c10/core/Device.h:3, 
Jan 31 21:59:04                  from /var/lib/jenkins/workspace/build/aten/src/ATen/core/TensorBody.h:3, 
Jan 31 21:59:04                  from /var/lib/jenkins/workspace/aten/src/ATen/core/jit_type.h:3, 
Jan 31 21:59:04                  from /var/lib/jenkins/workspace/torch/csrc/jit/script/class_type.cpp:1: 
Jan 31 21:59:04 /var/lib/jenkins/workspace/torch/csrc/jit/script/class_type.cpp: In member function 'const std::vector<torch::jit::Function*>& c10::ClassType::methods() const': 
Jan 31 21:59:04 /var/lib/jenkins/workspace/c10/util/Exception.h:215:29: error: nonnull argument 'this' compared to NULL [-Werror=nonnull-compare] 
Jan 31 21:59:04    if (C10_UNLIKELY_OR_CONST(!(cond))) {       \ 
Jan 31 21:59:04                              ^~~~~~~ 
Jan 31 21:59:04 /var/lib/jenkins/workspace/c10/macros/Macros.h:133:65: note: in definition of macro 'C10_UNLIKELY' 
Jan 31 21:59:04  #define C10_UNLIKELY(expr)  (__builtin_expect(static_cast<bool>(expr), 0)) 
Jan 31 21:59:04                                                                  ^~~~ 
Jan 31 21:59:04 /var/lib/jenkins/workspace/c10/util/Exception.h:215:7: note: in expansion of macro 'C10_UNLIKELY_OR_CONST' 
Jan 31 21:59:04    if (C10_UNLIKELY_OR_CONST(!(cond))) {       \ 
Jan 31 21:59:04        ^~~~~~~~~~~~~~~~~~~~~ 
Jan 31 21:59:04 /var/lib/jenkins/workspace/torch/csrc/jit/script/class_type.cpp:27:3: note: in expansion of macro 'TORCH_INTERNAL_ASSERT' 
Jan 31 21:59:04    TORCH_INTERNAL_ASSERT(this); 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[JIT] Fix python pickle serialization for torchbind"


Differential Revision: [D19669879](https://our.internmc.facebook.com/intern/diff/D19669879)

[ghstack-poisoned]

dzhulgakov reviewed (commented):
This is cool. The only question is it following the same format as torch.jit.save pickling? We're trying to make our JIT's torch.save and python's torch.save the same. I know it's contrived, but I bet someone will come up with the need to save something in jit and load it in python. It should be relatively easily doable if we just follow the same instructions format inside JIT's pickler

Commit:
Update on "[JIT] Fix python pickle serialization for torchbind"


Differential Revision: [D19669879](https://our.internmc.facebook.com/intern/diff/D19669879)

[ghstack-poisoned]

driazati commented on test/test_jit.py:
```diff
@@ -5058,6 +5058,17 @@ def forward(self):
 
+    @skipIfRocm
+    @unittest.skipIf(IS_WINDOWS, "TODO: Fix this test case")
+    def test_torchbind_pickle_serialization(self):
```
Can you also test that it is loadable as an attribute? Just add it to this list and run `TestScript.test_attribute_serialization`

https://github.com/pytorch/pytorch/blob/040bc1d0e1eac9c50072332fc4d939a90823099c/test/test_jit.py#L16513

driazati reviewed (approved):
This looks fine. Pybind11 should handle the pickle serialization which will save everything as the `ScriptObject` class but it should work when restored with plain `pickle`

Commit:
Update on "[JIT] Fix python pickle serialization for torchbind"


Differential Revision: [D19669879](https://our.internmc.facebook.com/intern/diff/D19669879)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix python pickle serialization for torchbind"


Differential Revision: [D19669879](https://our.internmc.facebook.com/intern/diff/D19669879)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix python pickle serialization for torchbind"


Differential Revision: [D19669879](https://our.internmc.facebook.com/intern/diff/D19669879)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix python pickle serialization for torchbind"


Differential Revision: [D19669879](https://our.internmc.facebook.com/intern/diff/D19669879)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix python pickle serialization for torchbind"


Differential Revision: [D19669879](https://our.internmc.facebook.com/intern/diff/D19669879)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix python pickle serialization for torchbind"


Differential Revision: [D19669879](https://our.internmc.facebook.com/intern/diff/D19669879)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix python pickle serialization for torchbind"


Differential Revision: [D19669879](https://our.internmc.facebook.com/intern/diff/D19669879)

[ghstack-poisoned]
